EIPS/eip-web3-access.md

```diff
@@ -0,0 +1,109 @@
+---
+eip: <to be assigned>
+title: Opt-in web3 access
+author: Paul Bouchon <mail@bitpshr.net>
+discussions-to: https://ethereum-magicians.org/t/opt-in-web3-access/414
+status: Draft
+type: Standards Track
+category: Interface
+created: 2018-05-04
+---
+
+## Simple summary
+
+This proposal describes a new way for environments to expose the web3 API that requires user approval.
+
+## Abstract
+
+MetaMask and most other tools that provide access to web3-enabled environments do so automatically and without user consent. This exposes users of such environments to fingerprinting attacks since untrusted websites can check for a `web3` object and reliably identify web3-enabled clients.
+
+This proposal outlines a new dapp initialization strategy in which websites request access to the web3 API instead of relying on its preexistence in a given environment.
+
+## Specification
+
+### Typical dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    STOP dapp
+```
+
+### Proposed dapp initialization
+
+```
+START dapp
+IF web3 is defined
+    CONTINUE dapp
+IF web3 is undefined
+    REQUEST[1] web3 API
+    IF user approves
+        INJECT[2] web3 API
+        NOTIFY[3] dapp
+        CONTINUE dapp
+    IF user rejects
+        STOP dapp
+```
+
+**REQUEST[1]** This operation would be achieved by an implementation-level messaging API like `window.postMessage` and should pass a payload containing a `type` property with a value of “WEB3_API_REQUEST”.
+
+**INJECT[2]** This operation would be achieved by any implementation-level API that can expose the web3 API to the user’s browser context, such as HTML script tag injection.
+
+**NOTIFY[3]** This operation would be achieved by an implementation-level messaging API like `window.postMessage` and should pass a payload containing a `type` property with a value of “WEB3_API_SUCCESS” after successful web3 exposure or “WEB3_API_ERROR” after unsuccessful web3 exposure. In the case of an error, an optional `message` property can be included with additional information.
```

Comment:
I believe this needs to be standardized, not implementation specific.  As a dapp author, having every web3 enabled browser notify me differently results in me living like a circa 2003 web developer.